collection_requirements: add requirements version specification recommendation to avoid conflicts when building EEs (#2585)
ansible/ansible-documentation: docs/docsite/rst/community/collection_contributors/collection_requirements.rst

meta/execution-environment.yml
------------------------------

If a collection has controller-side Python package and/or system package requirements, to allow easy :ref:`execution environment<getting_started_ee_index>` building:

* They SHOULD be listed in corresponding files under the ``meta`` directory, specified in ``meta/execution-environment.yml``, and `verified <https://ansible.readthedocs.io/projects/builder/en/latest/collection_metadata/#when-installing-collections-using-ansible-galaxy>`_.

> * The entries in the file SHOULD NOT have a version cap or be fixed. This means no ``<=x.x.x`` or ``==x.x.x`` entries, only ``>=x.x.x`` entries. No specified version at all is also allowed.

Reviewer: I would extend this to:
```suggestion
* The entries in the file SHOULD NOT have a version cap or be fixed. This means no ``<=x.x.x`` or ``==x.x.x`` entries, only ``>=x.x.x`` entries, potentially combined with ``,!=x.x.x``. No specified version at all is also allowed.
```

Author: LGTM, committed, thanks, PTAL

> * The entries in the file SHOULD NOT have a version cap or be fixed. This means no ``<=x.x.x`` or ``==x.x.x`` entries, only ``>=x.x.x`` entries, potentially combined with ``,!=x.x.x``. No specified version at all is also allowed.

Reviewer 2: ```suggestion
* The entries in the file SHOULD NOT have a version cap or be fixed. Entries like ``<=x.x.x`` or ``==x.x.x`` are not allowed. Only ``>=x.x.x`` entries (with or without ``,!=x.x.x``), or no specified version at all are allowed.
```

>   * It helps prevent issues during the creation of execution environments, particularly when bundling collections that have conflicting versions of the same dependencies. For instance, if one collection requires a fixed version of a dependency while another requires a higher version of the same dependency, the build process will fail.

Reviewer 3: Given [Russoz's comment](https://forum.ansible.com/t/vote-ends-on-2025-05-20-collection-requirements-dependencies-related-changes/42281/9) are there other risks and avoidance strategies we might want to document.

It's good to detail that WHY (rational) as it provides a lot more context to help the reader understand why these rules exist

Author: I think the only risk is the dependencies conflict, which we describe here I believe well

>   * It helps prevent issues during the creation of execution environments, particularly when bundling collections that have conflicting versions of the same dependencies. For instance, if one collection requires a fixed version of a dependency while another requires a higher version of the same dependency, the build process will fail.

Reviewer 2: ```suggestion
  * This rule helps prevent issues during the creation of execution environments, particularly when bundling collections that have conflicting versions of the same dependencies. For instance, if one collection requires a fixed version of a dependency while another requires a higher version of the same dependency, the build process will fail.
```
